File: Algorithms/Algorithms/Algorithms/Heap.cpp
#include "Heap.h"

int Parent(int i)
{
	if (i != 0) {
		return i / 2;
	}
}

int Left(std::vector<int>* vec, int i)
{
	if (2 * i <= (*vec).size()) {
		return 2 * i;
	}
	else {
		return i;
	}
}

int Right(std::vector<int>* vec , int i)
{
	if ((2 * i) + 1 <= (*vec).size()) {
		return (2 *i)+1;
	}
	else {
		return i;
	}
}

void insert(std::vector<int>* vec, int x)
{
	(*vec).resize((*vec).size() + 1);
	int i = (*vec).size()-1;
	(*vec)[i] = x;
	while (i > 0 && (*vec)[Parent(i)] < (*vec)[i])
	{
		int tmp = (*vec)[i];
		(*vec)[i] = (*vec)[Parent(i)];
		(*vec)[Parent(i)] = tmp;

		i = Parent(i);
	}
}

void extract(std::vector<int>* vec)
{
	(*vec)[0] = (*vec)[(*vec).size() - 1];
	(*vec).resize((*vec).size() - 1);
	heapify(vec, 0);
}

void heapify(std::vector<int>* vec, int i)
{
	int largest;

	if (Left(vec, i) <= (*vec).size() - 1 && (*vec)[Left(vec, i)] > (*vec)[i]) {
		largest = Left(vec, i);
	}
	else {
		largest = i;
	}

	if (Right(vec, i) <= (*vec).size() - 1 && (*vec)[Right(vec, i)] > (*vec)[largest]) {
		largest = Right(vec, i);
	}

	if (largest != i) {
		std::swap((*vec)[i], (*vec)[largest]);
		heapify(vec, largest);
	}
}

File: ASD_Projects/Progetto 4/Graph/src/Heap.java
import java.util.*;

public class Heap<T> implements AbstractHeap<T>{

    private int heap_size = 0;
    private int key = 0;
    T x;

    boolean MinMax= true; //true = max

    private ArrayList<Map.Entry<Integer,T>> heap;
    private Comparator<T> comparator;
    public int getHeap_size() {
        return heap_size;
    }
    public Heap(Comparator<T> comparator) {
        this.comparator = comparator;
        heap = new ArrayList<Map.Entry<Integer,T>>();
    }

    public ArrayList<Map.Entry<Integer, T>> getHeap() {
        return heap;
    }

    public void setMaxHeap(){
        this.MinMax = true;
    }

    public void setMinHeap(){
        this.MinMax = false;
    }

    private int getKey(){
        return ++key;
    }
    public Map.Entry<Integer, T> getFirst(){
            try {
                return this.heap.getFirst();
            }catch (NoSuchElementException e){
                System.out.println("Heap is empty");
            }
        return null;
    }

    public Map.Entry<Integer, T> get(int index){
        try {
            return heap.get(index);
        }catch (IndexOutOfBoundsException e){
            System.out.println("Heap does not reach " + index + "th element");
        }
       return null;
    }

    @Override
    public int Parent(int index) {
        return index/2;
    }

    @Override
    public int Left(int index){
        return index*2;
    };

    @Override
    public int Right(int index){
        return index*2+1;
    }

    @Override
    public AbstractMap.SimpleEntry<Integer,T> Insert(T value) {
        heap_size++;
        int i = heap_size-1;
        AbstractMap.SimpleEntry<Integer,T> am = new AbstractMap.SimpleEntry<Integer,T>(value.toString().hashCode(),value);
        heap.addLast(am);
        //select operation based on the type of heap desired
        if(this.MinMax){
            while (i > 0 && comparator.compare(heap.get(Parent(i)).getValue(), heap.get(i).getValue()) < 0) {
                Collections.swap(heap, i, Parent(i));
                i = Parent(i);
            }
        }else{
            while (i > 0 && comparator.compare(heap.get(Parent(i)).getValue(), heap.get(i).getValue()) > 0) {
                Collections.swap(heap, i, Parent(i));
                i = Parent(i);
            }
        }
           return am;
    }

    @Override
    public void Extract() {
        try{
            heap.removeFirst();
        }catch (NoSuchElementException e){
            System.out.println("Heap is empty");
        }

        heap_size--;
        Heapify(0);
    }

    public void Remove(int index){
        try{
            heap.remove(index);
        }catch (IndexOutOfBoundsException e){
            System.out.println("Heap does not reach " + index + "th element");
        }
        heap_size--;
        //Heapify(0); //todo: check if needs to uncomment this
    }

    @Override
    public void Heapify(int index) {
        if(this.MinMax){
            int largest = index;
            //find max
            if(comparator.compare(heap.get(Left(index)).getValue(), heap.get(index).getValue()) > 0){
                largest = Left(index);
            }

            if(comparator.compare(heap.get(Right(index)).getValue(), heap.get(index).getValue()) > 0){
                largest = Right(index);
            }
            if(largest != index){
                Collections.swap(heap,largest,index);
                Heapify(largest);
            }
        }else{
            int smallest = index;
            //find max
            if(comparator.compare(heap.get(Left(index)).getValue(), heap.get(index).getValue()) < 0){
                smallest = Left(index);
            }

            if(comparator.compare(heap.get(Right(index)).getValue(), heap.get(index).getValue()) < 0){
                smallest = Right(index);
            }
            if(smallest != index){
                Collections.swap(heap, smallest,index);
                Heapify(smallest);
            }
        }


    }

}
